test/perf/perf.test.js

```diff
@@ -0,0 +1,108 @@
+ * governing permissions and limitations under the License.
+ */
+/* eslint-disable no-undef */
+/* eslint-disable camelcase */
+const { fetch } = require('@adobe/helix-fetch');
+const { getUrls, parseTiming } = require('../utils');
+
+const baseService = process.env.HLX_FASTLY_NAMESPACE;
+const testService = process.env.TEST_SERVICE;
+/**
+ * generates strings with information about changes in rendering time
+ *
+ * @param {*} timeObj1 object of pipeline step times and desc
+ * @param {*} timeObj2 object of pipeline step times and desc
+ */
+function testTiming(timeObj1, timeObj2) {
+  const actual = Object.keys(timeObj2).reduce((prev, curr) => {
+    if (curr in timeObj1) {
+      const baseTime = timeObj1[curr];
+      const branchTime = timeObj2[curr];
+      const diff = branchTime - baseTime;
+      // eslint-disable-next-line no-param-reassign
+      prev += `commit will ${diff < 0 ? 'DECREASE' : 'INCREASE'} time of step: ${curr} by ${Number(diff < 0 ? (-1.0 * diff) : diff).toFixed(4)}ms\n`;
+    } else {
+      // eslint-disable-next-line no-param-reassign
+      prev += `commit adds step ${curr}: ${branchTime}ms\n`;
+    }
+    return prev;
+  }, '\n');
+
+  const expected = Object.keys(timeObj1).reduce((prev, curr) => {
+    if (!(curr in timeObj2)) {
+      const baseTime = timeObj1[curr];
+      // eslint-disable-next-line no-param-reassign
+      prev += `commit removes step ${curr}: ${baseTime}ms\n`;
+    }
+    return prev;
+  }, actual);
+
+  return { actual, expected };
+}
+
+/**
+ * constructs list of promises of calls to fetch
+ *
+ */
+async function sendPerfRequests() {
+  const urls = await getUrls();
+  const baseHeaders = {
+    'X-DEBUG': baseService,
+    'cache-control': 'no-cache',
+  };
+  const testHeaders = {
+    'X-DEBUG': testService,
+    'cache-control': 'no-cache',
+  };
+  const visits = urls.reduce((prev, curr) => {
+    prev.push(fetch(`${curr.base}`, { headers: baseHeaders }).then((res) => {
+      // consume response
+      res.text();
+      return res;
+    }));
+    prev.push(fetch(`${curr.branch}`, { headers: testHeaders }).then((res) => {
+      // consume response
+      res.text();
+      return res;
+    }));
+    return prev;
+  }, []);
+
+  return Promise.all(visits);
+}
+/**
+ * runs performance tests and compares metrics
+ *
+ */
+async function runPerfTests(res) {
+  describe('pipeline rendering time', () => {
+    for (let idx = 0; idx < res.length; idx += 2) {
+      const { url } = res[idx];
+      const baseTime = res[idx].headers.get('server-timing');
+      const branchTime = res[idx + 1].headers.get('server-timing');
+      if (baseTime && branchTime) {
+        const baseTimeObj = parseTiming(baseTime);
+        const branchTimeObj = parseTiming(branchTime);
+        it(`testing pipeline rendering time for ${url}`, () => {
+          const { actual } = testTiming(baseTimeObj, branchTimeObj);
+          // eslint-disable-next-line no-console
+          console.log(actual);
```

Comment:
If we output only `actual`, how do we know the diff ? Maybe we should output `actual` and `expected` with a user friendly comment ?
Maybe we can also `console.warn` / ` console.error` is the diff is bigger than a threshold ?
I really wonder what those numbers will tell us!

Reply:
If it exceeds a threshold (probably a percentage) the test should fail. Without an actual assertion, this isn't a real test.

Reply:
^^ Right; what do you think is a good percentage?

Reply:
I think I have an idea on how we can construct the percentage change that counts as significant! I can create a script that'll just send out a ton of requests to helix sites. We can then just determine the distribution of the rendering times; and then we can determine what it means for a rendering time to be an outlier! @kptdobe @trieloff WDYT

Maybe we can create a kind of perf config file? that contains info on pipeline execution metrics. I think this can turn into a larger discussion about monitoring pipeline performance in general!

Reply:
> Without an actual assertion, this isn't a real test.
I would prefer this to not become a test yet. I already observed in the Helix Pages integration tests that sometimes they timeout after 5s because the request (to www.hlx.page for example) takes too long to come back. I think observing the timing is a good first step because I am almost sure we will not be able to rely on this metric: it is not stable!